private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,280 @@
+func (l *lsp) init(params *protocol.InitializeParams) error {
+	if l.initParams.Load() != nil {
+		return fmt.Errorf("called the %q method more than once", protocol.MethodInitialize)
+	}
+	l.initParams.Store(params)
+
+	// TODO: set up logging. We need to forward everything from server.logger through to
+	// the client, if tracing is turned on. The right way to do this is with an extra
+	// goroutine and some channels.
+
+	return nil
+}
+
+// findImportable finds all files that can potentially be imported by the proto file at
+// uri. This returns a map from potential Protobuf import path to the URI of the file it would import.
+//
+// Note that this performs no validation on these files, because those files might be open in the
+// editor and might contain invalid syntax at the moment. We only want to get their paths and nothing
+// more.
+func (l *lsp) findImportable(
+	ctx context.Context,
+	uri protocol.URI,
+) (map[string]bufimage.ImageFileInfo, error) {
+	fileInfos, err := l.controller.GetImportableImageFileInfos(ctx, uri.Filename())
+	if err != nil {
+		return nil, err
+	}
+
+	imports := make(map[string]bufimage.ImageFileInfo)
+	for _, fileInfo := range fileInfos {
+		imports[fileInfo.Path()] = fileInfo
+	}
+
+	l.logger.Sugar().Debugf("found imports for %q: %#v", uri, imports)
+
+	return imports, nil
+}
+
+// newHandler constructs an RPC handler that wraps the default one from jsonrpc2. This allows us
+// to inject debug logging, tracing, and timeouts to requests.
+func (l *lsp) newHandler() jsonrpc2.Handler {
+	actual := protocol.ServerHandler(newServer(l), nil)
+	return func(ctx context.Context, reply jsonrpc2.Replier, req jsonrpc2.Request) (retErr error) {
+		ctx, span := l.tracer.Start(
+			ctx,
+			tracing.WithErr(&retErr),
+			tracing.WithAttributes(attribute.String("method", req.Method())),
+		)
+		defer span.End()
+
+		l.logger.Debug(
+			"processing request",
+			zap.String("method", req.Method()),
+			zap.ByteString("params", req.Params()),
+		)
+
+		// Each request is handled in a separate goroutine, and has a fixed timeout.
+		// This is to enforce responsiveness on the LSP side: if something is going to take
+		// a long time, it should be offloaded.
+		ctx, done := context.WithTimeout(ctx, handlerTimeout)
+		ctx = withRequestID(ctx)
+
+		go func() {
+			defer done()
+			replier := l.adaptReplier(reply, req)
+
+			// Verify that the server has been initialized if this isn't the initialization
+			// request.
+			if req.Method() != protocol.MethodInitialize && l.initParams.Load() == nil {
+				retErr = replier(ctx, nil, fmt.Errorf("the first call to the server must be the %q method",
+					protocol.MethodInitialize))
+				return
+			}
+
+			retErr = actual(ctx, replier, req)
+		}()
+
+		<-ctx.Done()
+		if ctx.Err() == context.DeadlineExceeded {
+			// Don't return this error; that will kill the whole server!
+			l.logger.Sugar().Errorf("timed out while handling %s; this is likely a bug", req.Method())
+		}
+		return retErr
```

Comment:
This looks like a data race on `retErr`. I'm not sure you need the go routine here as passing down the context should be enough to ensure the work is bounded.

Reply:
No, this is necessary because jsonrpc2 does not parallelize requests by default.

Reply:
This go routine doesn't parallelize the processing of requests as it blocks and waits for the completion. The same blocking can be done without the go routine and rely on the context to cancel the work. There still looks like a race condition on retErr as the above return on context cancel doesn't guarantee the write happens before the read.

Reply:
It waits for completion or deadline expiration. Your observation re: relying on the context to cancel the work is only correct if the work is I/O bound, but that is not the case here.

There is no data race; I suggest looking at how context cancelation and synchronization works (writing to a channel is synchronization).

Reply:
It's an unlikely data race, but the context can be canceled by the caller without `context.DeadlineExceeded ` so  it's still racy. CPU bounded can still take ctx and check to close work. Guess for a short lived CPU only this is fine, but most functions should take ctx and check it so it doesn't have to be only I/O bound work.

Reply:
I think this could be written with an explicit channel.
```go
errChan := make(chan err, 1)
go func() { errChan <- work(ctx) }()
select {
case err <- errChan:
    return err
case <- ctx.Done():
    return ctx.Err()
}
```